File: apps/ios/Sources/ViewModels/TriageViewModel.swift
import Foundation

// MARK: - TriageStatus

/// Conversion status for triage reports.
enum TriageStatusFilter: String, CaseIterable, Sendable {
    case all
    case pending
    case reviewing
    case converted
    case dismissed
    case inProgress = "in_progress"
    case completed

    var displayName: String {
        switch self {
        case .all: return NSLocalizedString("triage_filter_all", comment: "All")
        case .pending: return NSLocalizedString("triage_status_pending", comment: "Pending")
        case .reviewing: return NSLocalizedString("triage_status_reviewing", comment: "Reviewing")
        case .converted: return NSLocalizedString("triage_status_converted", comment: "Converted")
        case .dismissed: return NSLocalizedString("triage_status_dismissed", comment: "Dismissed")
        case .inProgress: return NSLocalizedString("triage_filter_in_progress", comment: "In Progress")
        case .completed: return NSLocalizedString("triage_filter_completed", comment: "Completed")
        }
    }
}

// MARK: - TriageViewModel

/// Manages the triage queue: reports with `allowCaseConversion: true` that can be
/// converted to case records. Fetches from `GET /api/reports?conversionEnabled=true`.
@Observable
final class TriageViewModel {
    private let apiService: APIService
    private let cryptoService: CryptoService

    var reports: [ClientReportResponse] = []
    var total: Int = 0
    var isLoading = false
    var isActionInProgress = false
    var errorMessage: String?

    /// Current conversion status filter.
    var selectedFilter: TriageStatusFilter = .pending

    /// Filtered reports based on selected conversion status.
    var filteredReports: [ClientReportResponse] {
        guard selectedFilter != .all else { return reports }
        // The conversionStatus filter is applied server-side, but for client-side
        // filtering of already-loaded data:
        return reports
    }

    /// Report type definitions for resolving labels.
    var reportTypes: [ClientReportTypeDefinition] = []

    init(apiService: APIService, cryptoService: CryptoService) {
        self.apiService = apiService
        self.cryptoService = cryptoService
    }

    // MARK: - Load Triage Reports

    func loadReports() async {
        guard !isLoading else { return }
        isLoading = true
        errorMessage = nil

        async let reportsResult: Void = fetchReports()
        async let typesResult: Void = fetchReportTypes()

        await reportsResult
        await typesResult

        isLoading = false
    }

    private func fetchReports() async {
        do {
            var path = apiService.hp("/api/reports") + "?conversionEnabled=true&limit=50"
            if selectedFilter != .all {
                path += "&conversionStatus=\(selectedFilter.rawValue)"
            }
            let response: ReportsListResponse = try await apiService.request(
                method: "GET",
                path: path
            )
            reports = response.conversations
            total = response.total
        } catch {
            if reports.isEmpty {
                errorMessage = error.localizedDescription
            }
        }
    }

    private func fetchReportTypes() async {
        do {
            let types = try await apiService.fetchCmsReportTypes()
            reportTypes = types
        } catch {
            // Report types are optional
        }
    }

    // MARK: - Filter

    func filterByStatus(_ status: TriageStatusFilter) async {
        selectedFilter = status
        reports = []
        total = 0
        await fetchReports()
    }

    // MARK: - Refresh

    func refresh() async {
        isLoading = false
        await loadReports()
    }

    // MARK: - Convert to Case

    /// Convert a triage report to a case record.
    ///
    /// Convert a triage report to a full entity record using the atomic conversion endpoint.
    ///
    /// - Parameters:
    ///   - report: The report to convert.
    ///   - entityTypeId: The target entity type ID selected by the user.
    /// - Returns: `true` if conversion succeeded.
    @discardableResult
    func convertToEntity(report: ClientReportResponse, entityTypeId: String) async -> Bool {
        isActionInProgress = true
        errorMessage = nil

        do {
            let body = AppConvertFromReportBody(
                reportId: report.id,
                entityTypeId: entityTypeId,
                additionalFields: nil
            )
            let _: AppConvertFromReportResponse = try await apiService.request(
                method: "POST",
                path: apiService.hp("/api/records/convert-from-report"),
                body: body
            )
            await refresh()
            isActionInProgress = false
            return true
        } catch {
            errorMessage = error.localizedDescription
            isActionInProgress = false
            return false
        }
    }

    // MARK: - Helpers

    func reportTypeLabel(for typeId: String?) -> String? {
        guard let typeId else { return nil }
        return reportTypes.first { $0.id == typeId }?.label
    }
}

// MARK: - Request/Response Types

struct AppConvertFromReportBody: Encodable, Sendable {
    let reportId: String
    let entityTypeId: String
    let additionalFields: [String: String]?
}

struct AppConvertFromReportResponse: Codable, Sendable {
    let recordId: String
    let reportId: String
    let entityTypeId: String
    let caseNumber: String?
    let autoAssigned: Bool
    let assignedTo: [String]
}

File: apps/android/app/src/main/java/org/llamenos/hotline/ui/triage/TriageViewModel.kt
package org.llamenos.hotline.ui.triage

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.llamenos.hotline.api.ApiService
import org.llamenos.hotline.hub.ActiveHubState
import org.llamenos.hotline.model.CmsReportTypesResponse
import org.llamenos.hotline.model.ConvertReportToCaseRequest
import org.llamenos.hotline.model.ConvertReportToCaseResponse
import org.llamenos.hotline.model.Report
import org.llamenos.hotline.model.ReportsListResponse
import org.llamenos.protocol.ReportTypeDefinition
import javax.inject.Inject

/**
 * Conversion status filters for the triage queue.
 */
enum class TriageStatusFilter(val apiValue: String?) {
    ALL(null),
    PENDING("pending"),
    IN_PROGRESS("in_progress"),
    COMPLETED("completed"),
}

data class TriageUiState(
    val reports: List<Report> = emptyList(),
    val total: Int = 0,
    val isLoading: Boolean = false,
    val isRefreshing: Boolean = false,
    val error: String? = null,
    val selectedFilter: TriageStatusFilter = TriageStatusFilter.PENDING,
    val reportTypes: List<ReportTypeDefinition> = emptyList(),
    val isConverting: Boolean = false,
    val selectedReport: Report? = null,
)

/**
 * ViewModel for the triage queue screen.
 *
 * Fetches reports with `conversionEnabled=true` from the reports API.
 * Supports filtering by conversion status and converting reports to case records.
 */
@HiltViewModel
class TriageViewModel @Inject constructor(
    private val apiService: ApiService,
    private val activeHubState: ActiveHubState,
) : ViewModel() {

    private val _uiState = MutableStateFlow(TriageUiState())
    val uiState: StateFlow<TriageUiState> = _uiState.asStateFlow()

    init {
        activeHubState.activeHubId
            .filterNotNull()
            .onEach { loadTriageQueue() }
            .launchIn(viewModelScope)
    }

    fun loadTriageQueue() {
        viewModelScope.launch {
            _uiState.update {
                it.copy(
                    isLoading = it.reports.isEmpty(),
                    isRefreshing = it.reports.isNotEmpty(),
                    error = null,
                )
            }
            val reportsDeferred = async { fetchReports() }
            val typesDeferred = async { fetchReportTypes() }
            reportsDeferred.await()
            typesDeferred.await()
        }
    }

    private suspend fun fetchReports() {
        try {
            val filter = _uiState.value.selectedFilter
            val path = buildString {
                append(apiService.hp("/api/reports"))
                append("?conversionEnabled=true&limit=50")
                filter.apiValue?.let { append("&conversionStatus=$it") }
            }
            val response = apiService.request<ReportsListResponse>("GET", path)
            _uiState.update {
                it.copy(
                    reports = response.conversations,
                    total = response.total,
                    isLoading = false,
                    isRefreshing = false,
                )
            }
        } catch (e: Exception) {
            _uiState.update {
                it.copy(
                    isLoading = false,
                    isRefreshing = false,
                    error = e.message ?: "Failed to load triage queue",
                )
            }
        }
    }

    private suspend fun fetchReportTypes() {
        try {
            val response = apiService.request<CmsReportTypesResponse>(
                "GET",
                "/api/settings/cms/report-types",
            )
            _uiState.update { it.copy(reportTypes = response.reportTypes) }
        } catch (_: Exception) {
            // Report types are optional
        }
    }

    fun setFilter(filter: TriageStatusFilter) {
        _uiState.update { it.copy(selectedFilter = filter, reports = emptyList(), total = 0) }
        loadTriageQueue()
    }

    fun refresh() {
        loadTriageQueue()
    }

    fun selectReport(report: Report) {
        _uiState.update { it.copy(selectedReport = report) }
    }

    fun dismissError() {
        _uiState.update { it.copy(error = null) }
    }

    /**
     * Convert a report to a case record.
     */
    fun convertToCase(report: Report) {
        viewModelScope.launch {
            _uiState.update { it.copy(isConverting = true, error = null) }
            try {
                val title = report.metadata?.reportTitle ?: "Untitled Report"
                val body = ConvertReportToCaseRequest(
                    reportId = report.id,
                    title = title,
                    reportTypeId = report.metadata?.reportTypeId,
                )
                apiService.request<ConvertReportToCaseResponse>(
                    "POST",
                    apiService.hp("/api/reports/${report.id}/convert-to-case"),
                    body = body,
                )
                _uiState.update { it.copy(isConverting = false, selectedReport = null) }
                // Reload the queue
                loadTriageQueue()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isConverting = false,
                        error = e.message ?: "Failed to convert report",
                    )
                }
            }
        }
    }

    /**
     * Resolve a report type label from its ID.
     */
    fun reportTypeLabel(typeId: String?): String? {
        if (typeId == null) return null
        return _uiState.value.reportTypes.find { it.id == typeId }?.label
    }
}
